File: 1155-number-of-dice-rolls-with-target-sum/1155-number-of-dice-rolls-with-target-sum.cpp
const int MOD = 1e9 + 7;

class Solution {
public:
    long long dp[35][1005];
    
    int numDice;
    int numFaces;
    int target;
    
    int solve(int n, int sum) {
        if (n == 0) {
            return sum == target;
        }
        if (sum < 0 || sum > target) {
            return 0;
        }
        if (dp[n][sum] != -1) {
            return dp[n][sum];
        }
        long long res = 0;
        for (int i = 1; i <= numFaces; i++) {
            res = (res + solve(n - 1, sum + i)) % MOD;
        }
        return dp[n][sum] = res;
    }
    
    int numRollsToTarget(int n, int k, int sum) {
        target = sum;
        numDice = n;
        numFaces = k;
        memset(dp, -1, sizeof(dp));
        return solve(n, 0);
    }
};

File: 1155-number-of-dice-rolls-with-target-sum/1155-number-of-dice-rolls-with-target-sum.java
class Solution {
    public int numRollsToTarget(int n, int k, int target) {
        long[][] dp = new long[n + 1][target + 1];
        dp[0][0] = 1;
        int MOD = (int) 1e9 + 7;
        for (int i = 1; i <= n; i++) {
            for (int j = 1; j <= target; j++) {
                for (int face = 1; face <= k && face <= j; face++) {
                    dp[i][j] = (dp[i][j] + dp[i - 1][j - face]) % MOD;
                }
            }
        }
        return (int) dp[n][target];
    }
}